python/cudf/cudf/core/_base_index.py

```diff
@@ -1414,6 +1423,128 @@ def from_pandas(cls, index, nan_as_null=None):
     def _constructor_expanddim(self):
         return cudf.MultiIndex
 
+    def drop_duplicates(
+        self, keep="first", nulls_are_equal=True,
+    ):
+        """
+        Drop duplicate rows in index.
+
+        keep : ["first", "last", False]
+            "first" will keep first of duplicate, "last" will keep last of the
+            duplicate and "False" drop all duplicate.
+        nulls_are_equal: bool, default True
+            Null elements are considered equal to other null elements.
+        """
+
+        # This utilizes the fact that all `Index` is also a `Frame`.
+        result = self.__class__._from_columns(
+            drop_duplicates(
+                list(self._columns),
+                keys=range(len(self._data)),
+                keep=keep,
+                nulls_are_equal=nulls_are_equal,
+            ),
+            self._column_names,
+        )
+        result._copy_type_metadata(self)
+        return result
+
+    def dropna(self, how="any"):
+        """
+        Drop null rows from Index.
+
+        how : {"any", "all"}, optional
+            Specifies how to decide whether to drop a row.
+            "any" (default) drops rows containing at least
+            one null value. "all" drops only rows containing
+            *all* null values.
+        """
+
+        # This is to be consistent with IndexedFrame.dropna to handle nans
+        # as nulls by default
+        data_columns = [
+            col.nans_to_nulls()
+            if isinstance(col, cudf.core.column.NumericalColumn)
+            else col
+            for col in self._columns
+        ]
+
+        result = self.__class__._from_columns(
+            drop_nulls(data_columns, how=how, keys=range(len(data_columns)),),
+            self._column_names,
+        )
+        result._copy_type_metadata(self)
+        return result
+
+    def _gather(self, gather_map, nullify=False, check_bounds=True):
+        """Gather rows of index specified by indices in `gather_map`.
+
+        Skip bounds checking if check_bounds is False.
+        Set rows to null for all out of bound indices if nullify is `True`.
+        """
+        gather_map = cudf.core.column.as_column(gather_map)
+
+        # TODO: For performance, the check and conversion of gather map should
+        # be done by the caller. This check will be removed in future release.
+        if not is_integer_dtype(gather_map.dtype):
+            gather_map = gather_map.astype("int32")
+
+        if not _gather_map_is_valid(
```

Comment:
Not something that I noticed before, but the name here is a little too broad for my taste. Since the only validity being checked is whether or not the indices are in bounds, a more specific name might be better for the helper function. That change isn't necessary in this PR, just something to consider for a follow-up

Reply:
I guess another reason that makes it broad is for all these arguments "check_bounds", "nullify", it's kinda hard to find a fit-all name. If we rename this to be something specific, like `_check_gather_map_bounds`, then `nullify` and `check_bounds` argument shouldn't be passed to this function. As

```python
if not nullify and check_bounds and _check_gather_map_bounds(indices, nrows)
```
is more expressive to
```python
if _check_gather_map_bounds(indices, nrows, check_bounds, nullify)
```

Reply:
That's fair, I agree that your first snippet is more readable if you change the name but then it forces you to write more verbose code everywhere that does the check.

Reply:
Should this utility live in `lib.copying` instead of `utils`?

Reply:
@shwina somehow I missed this comment :/ Why keep it in the cython layer if only python codes call it?

Reply:
From offline that we are going to move this helper into `lib.copying` to restrict usage since it's specific to gather operation.